### PR TITLE
Reuse lookahead frame state and motion vectors

### LINKED ISSUE #1677: Reuse motion search data from lookahead
Lookahead does qres/hres motion search that is nearly identical to the qres/hres motion search done at the beginning of encoding a frame - the only difference being that lookahead uses lossless reference frames rather than the encoded one. It may be possible to skip the qres/hres motion search at the beginning of frame encoding and just reuse the lookahead results.

### PR DESCRIPTION
**Pull request by shssoichiro**
Results in 5-8% encoding time improvement with 1 tile. [AWCY](https://beta.arewecompressedyet.com/?job=master-edbc586999d6a1e49e1e35d84a0c8a2c9dcf7990&job=reuse-lookahead-mvs%402019-11-24T06%3A32%3A54.455Z)

Closes #1677

### CODE REVIEW

**Comment by easyfab** [comment]
I think something is broken with tiles. can someone confirm ?
 e.g for 1080p --tiles 4 encoding is ok but  with --tiles 8  rav1e exit at start without error message.
tested on windows 10 with appveyor build 1.0.4077

**Comment by shssoichiro** [comment]
With the last set of changes, I can no longer reproduce crashes with the latest rustc, but CI is consistently segfaulting on rustc 1.36, which is a real wtf. Going to dig in a bit locally to see what the heck is going on.

**Comment by shssoichiro** [comment]
Okay, this _should_ be working now. I ran a variety of tests, and Travis is passing now. AWCY and speed change remains the same as initially. Given how this PR has gone, I would like to get at least one other person to give it a test and verify that it no longer segfaults.

File: src/encoder.rs [hunk 1]
```diff
@@ -3051,7 +3080,6 @@ pub(crate) fn build_half_res_pmvs<T: Pixel>(
 pub(crate) fn build_full_res_pmvs<T: Pixel>(
   fi: &FrameInvariants<T>, ts: &mut TileStateMut<'_, T>,
   tile_sbo: TileSuperBlockOffset,
-  half_res_pmvs: &[[[Option<MotionVector>; REF_FRAMES]; 5]],
```
Comment (lu-zero) [inline src/encoder.rs]: I would not do this, at least in this set.

File: src/rdo.rs [hunk 2]
```diff
@@ -1012,7 +1013,7 @@ fn inter_frame_rdo_mode_decision<T: Pixel>(
     if !fi.config.speed_settings.encode_bottomup
       && (bsize == BlockSize::BLOCK_32X32 || bsize == BlockSize::BLOCK_64X64)
     {
-      pmvs[ref_slot] = Some(b_me);
+      ts.half_res_pmvs[pmv_idxs.0][pmv_idxs.1][ref_slot] = Some(b_me);
```
Comment (lu-zero) [inline src/rdo.rs]: Why?

File: src/api/internal.rs [hunk 3]
```diff
@@ -619,7 +619,7 @@ impl<T: Pixel> ContextInner<T> {
               TileSuperBlockOffset(SuperBlockOffset { x: sbx, y: sby });
             let half_res_pmvs =
               build_half_res_pmvs(fi, ts, tile_sbo, &tile_pmvs);
-            ts.half_res_pmvs.push(half_res_pmvs);
+            ts.half_res_pmvs[sbx + sby * ts.sb_width] = half_res_pmvs;
```
Comment (lu-zero) [inline src/api/internal.rs]: To keep this safer would be a good idea to use the rayon collect capabilities to save the half_res_pmvs outside of the parallel loop.

File: src/tiling/tile_state.rs [hunk 4]
```diff
@@ -89,7 +89,8 @@ impl<'a, T: Pixel> TileStateMut<'a, T> {
     let sb_width = width.align_power_of_two_and_shift(sb_size_log2);
     let sb_height = height.align_power_of_two_and_shift(sb_size_log2);
     if !fs.half_res_pmvs.iter().any(|&(key, _)| key == sbo) {
-      fs.half_res_pmvs.push((sbo, Vec::with_capacity(sb_width * sb_height)));
+      fs.half_res_pmvs
+        .push((sbo, vec![Default::default(); sb_width * sb_height]));
```
Comment (lu-zero) [inline src/tiling/tile_state.rs]: Why?

File: src/api/internal.rs [hunk 5]
```diff
@@ -613,14 +613,13 @@ impl<T: Pixel> ContextInner<T> {
         let tile_pmvs = build_coarse_pmvs(fi, ts, &inter_cfg);
 
         // Compute the half-resolution motion vectors.
-        let mut half_res_pmvs = Vec::with_capacity(ts.sb_height * ts.sb_width);
```
[the remark this answers is not in the document]
Comment (shssoichiro) [reply src/api/internal.rs]: Oh. Yeah. That would be better. A second set of eyes does help. ^^;

File: src/api/internal.rs [hunk 6]
```diff
@@ -613,14 +613,13 @@ impl<T: Pixel> ContextInner<T> {
         let tile_pmvs = build_coarse_pmvs(fi, ts, &inter_cfg);
```
[the remark this answers is not in the document]
Comment (shssoichiro) [reply src/api/internal.rs]: I'm not sure that we need to. We only ever use the coarse PMVs for building the half res PMVs.

File: src/rdo.rs (hunk 2 above)
Comment (shssoichiro) [reply src/rdo.rs]: tl;dr: Borrow checker :cry:

File: src/api/internal.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (lu-zero) [reply src/api/internal.rs]: I see :)

**Comment by lu-zero** [comment]
Looks good beside the need of a squash.

**Comment by shssoichiro** [comment]
Will use Squash and Merge button. :slightly_smiling_face:

**Comment by lu-zero** [comment]
For the usual 1080p Bosphorus with 8 tiles:
```
 INFO  rav1e::api::internal > Using 8 tiles (4x2)
 INFO  rav1e::stats         > encoded 600 frames, 1.122 fps, 1955.96 Kb/s
```
```
 INFO  rav1e::api::internal > Using 8 tiles (4x2)
 INFO  rav1e::stats         > encoded 600 frames, 1.153 fps, 1954.70 Kb/s
```

A 2% speedup.

**Comment by shssoichiro** [comment]
Speedup is inverse to the number of tiles, because the single-threaded lookahead MV calculation still exists. Still, won't balk at 2% for 8 tiles.

**Pull request merged**